lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -845,15 +896,37 @@ static LogicalResult verifyConnectOp(ConnectOp connect) {
   FIRRTLType destType = connect.dest().getType().cast<FIRRTLType>();
   FIRRTLType srcType = connect.src().getType().cast<FIRRTLType>();
 
+  auto isPortOrInstancePort = [](Value a) -> bool {
+    auto op = a.getDefiningOp();
+    return !op || dyn_cast<InstanceOp>(op);
+  };
+
+  // If the source or destination is a port or instance port, then an optional
+  // outer flip, indicating the direction (input or output), should be stripped
+  // for type checking.
+  if (isPortOrInstancePort(connect.dest()))
+    if (auto destTypeFlip = destType.dyn_cast<FlipType>())
+      destType = destTypeFlip.getElementType();
+  if (isPortOrInstancePort(connect.src()))
+    if (auto srcTypeFlip = srcType.dyn_cast<FlipType>())
+      srcType = srcTypeFlip.getElementType();
+
+  // If the source is an invalid value, then only compare using passive types.
+  if (auto op = connect.src().getDefiningOp()) {
+    if (dyn_cast<InvalidValuePrimOp>(op)) {
+      destType = destType.getPassiveType();
+      srcType = srcType.getPassiveType();
```

Comment:
I'm not sure why invalid's are a special case here.  Is this because our lowering from the invalidate statement to invalid+connect is wrong?  It would be great to elimiante this special case - it will make the IR more regular and less likely to have bugs.

Reply:
Ah, good catch. I think this can be fully removed given that this PR is (after this code was added) now expanding `is invalid` to expanded connects. This weakening of type equivalence stemmed from me fixing tests before fixing the invalid handling in the parser.